File: go/p30.go
package main

import "fmt"

func basicPow(n, pow int) int {
    product := 1

    for i := 0; i != pow; i++ {
        product *= n
    }

    return product
}

func digitPowSum(n, pow int) int {
    sum := 0
    for n > 0 {
        sum += basicPow(n % 10, pow)
        n /= 10
    }
    return sum
}

func main() {
    sum := 0
    for n := 2; n != 295245; n++ {
        if n == digitPowSum(n, 5) {
            sum += n
        }
    }
    fmt.Println(sum)
}

File: cpp/p30.cpp
#include <iostream>
using namespace std;

int Digit5thPowerSum(int x)
{
    int d, d2;
    int sum = 0;
    while (x != 0)
    {
        d = x % 10;
        x /= 10;
        
        d2 = d * d;
        sum += d2 * d2 * d;
    }
    
    return sum;
}

int main()
{
    int sum = 0;
    for (int i = 2; i != 354294; ++i)
        if (Digit5thPowerSum(i) == i)
            sum += i;
    
    cout << sum << endl;
    
    return 0;
}
